src/els_compiler_diagnostics.erl

```diff
@@ -231,42 +256,59 @@ string_to_term(Value) ->
   end.
 
 -spec compile_file(string(), [atom()]) ->
-        {ok | error, [compiler_msg()], [compiler_msg()]}.
+        {{ok | error, [compiler_msg()], [compiler_msg()]}
+        , [els_diagnostics:diagnostic()]}.
 compile_file(Path, Dependencies) ->
   %% Load dependencies required for the compilation
-  Olds = [load_dependency(Dependency) || Dependency <- Dependencies],
-  Opts = lists:append([ macro_options()
-                      , include_options()
-                      , [ return_warnings
-                        , return_errors
-                        , basic_validation
-                        ]
-                      ]),
-  Res = compile:file(Path, Opts),
+  Olds = [load_dependency(Dependency, Path) || Dependency <- Dependencies],
+  Res = compile:file(Path, diagnostics_options()),
   %% Restore things after compilation
   [code:load_binary(Dependency, Filename, Binary)
-   || {Dependency, Binary, Filename} <- Olds],
-  Res.
+   || {{Dependency, Binary, Filename}, _} <- Olds],
+  Diagnostics = lists:flatten([ Diags || {_, Diags} <- Olds ]),
+  {Res, Diagnostics}.
 
 %% @doc Load a dependency, return the old version of the code (if any),
 %% so it can be restored.
--spec load_dependency(atom()) -> {atom(), binary(), file:filename()} | error.
-load_dependency(Module) ->
+-spec load_dependency(atom(), string()) ->
+        {{atom(), binary(), file:filename()}, [els_diagnostics:diagnostic()]}
+          | error.
+load_dependency(Module, IncludingPath) ->
   Old = code:get_object_code(Module),
-  case els_utils:find_module(Module) of
-    {ok, Uri} ->
-      Path = els_utils:to_list(els_uri:path(Uri)),
-      Opts = lists:append([macro_options(), include_options(), [binary]]),
-      case compile:file(Path, Opts) of
-        {ok, Module, Binary} ->
-          code:load_binary(Module, atom_to_list(Module), Binary);
-        Error ->
-          lager:warning("Error compiling dependency [error=~w]", [Error])
-      end;
-    {error, Error} ->
-      lager:warning("Error finding dependency [error=~w]", [Error])
-  end,
-  Old.
+  Diagnostics =
+    case els_utils:find_module(Module) of
+      {ok, Uri} ->
+        Path = els_utils:to_list(els_uri:path(Uri)),
+        Opts = lists:append([macro_options(), include_options(), [binary]]),
+        case compile:file(Path, Opts) of
+          {ok, Module, Binary} ->
+            code:load_binary(Module, atom_to_list(Module), Binary),
+            [];
+          Error ->
+            lager:warning("Error compiling dependency [error=~w]", [Error]),
+            %% Compile again to generate diagnostics
+            get_dependency_diagnostics(Uri, IncludingPath)
```

Comment:
Why compiling it again? We just did that. Shouldn't we add the options to the `compile:file` above so that we actually return warnings and errors?

Reply:
It seems I misread the docs, thought you did not get the warnings when specifying `binary` as an option.